File: Hex2Words.rb
#!ruby.exe

# AUTHOR: Robert DiCicco
# DATE: 2022-05-24
# Challenge 166 Hexadecimal Words ( Ruby )

letters =  { 'a'=>"a",'b'=>"b",'c'=>"c",'d'=>"d",'e'=>"e",'f'=>"f",'o'=>"0",'l'=>"1",'i'=>"1",'s'=>"5",'t'=>"7" };
selection = 0;

for selection in ARGV 
  selection = selection.to_i
  if (selection < 2  or selection > 8 ) 
    puts "Selection must be >= 2 and <= 8"
    exit
  end
end

File.foreach("Words.txt") { |line| 
  line = line.lstrip.rstrip.downcase 
  if line.size != selection 
    next
  end
  @arr = line.split("")
  found = 0
  for letter in @arr 
    if letters[letter] 
      found = found + 1 
      if (found == selection)
        line = line.tr('olist','01157')
        print "#{line} "
        STDOUT.flush
      end
    end
  end
}

# -----------------------------------------------------------------------------------------
# SAMPLE OUTPUT
# PS G:\Projects\Perl\Challenges> ruby .\Hex2Words.rb 8
# da7aba5e de11ca7e de7a11ed d1abe7e5 d15ab1ed d15c105e f007ba11 1501a7ed 0ff1c1a1 7ac71ca1 

# PS G:\Projects\Perl\Challenges> ruby .\Hex2Words.rb 7
# be51de5 c1a55e5 c1a551c c011ec7 dec1ded def1c17 d15ea5e 

# PS G:\Projects\Perl\Challenges> ruby .\Hex2Words.rb 6
# acce55 affec7 a55e55 a55157 ba771e be11ef b0771e ca571e c105ed c0ffee deba7e decade dec1de 
# defea7 de7a11 de7ec7 effec7 e57a7e fa11ed f15ca1 175e1f 1a7e57 11771e 0ff1ce 0ff5e7 5e1ec7 
# 5e771e 50c1a1 57ab1e

File: Hex2Words.pl
#!perl.exe
use strict;
use warnings;
use feature qw/say/;

#AUTHOR: Robert DiCicco
# DATE: 2022-05-23
# Challenge 166 Hexadecimal Words ( Perl )

my %letters = ( 'a'=>'a','b'=>'b','c'=>'c','d'=>'d','e'=>'e','f'=>'f','o'=>0,'l'=>1,'i'=>1,'s'=>5,'t'=>7);
my $found = 0;
my $selection = $ARGV[0];

die "Length Selection Error. Must be between 2 and 8 inclusive" if ($selection < 2 or $selection > 8);

open(FH, '<', "./Words.txt") or die "Can't open Words.txt\n";

while(my $word = <FH>) {
  $word = lc($word);
  $word =~ s/^\s+|\s+$//g;                  # remove both leading and trailing spaces
  next if (length($word) != $selection);    # only look at the slected length words
  my @arr = split(//,$word);                # create array of chars
  foreach my $val (@arr){                   # if we have a character not in out list, go to next
    if(! defined($letters{$val})){
      next;
    }
    if(exists($letters{$val})){             # if its in our list, bump the found flag
      $found++;
    }
  }
  $word =~ tr/olist/01157/;       
  
  if ($found == $selection) {       # convert certain chars to their number representations{
    print "$word ";
  }

     $found = 0;                             # reset the found flag, go another round
  }
  print "\n";

#---------------------------------------------------------------------------
# SAMPLE OUTPUT
# PS G:\Projects\Perl\Challenges> perl .\Hex2Words.pl 8
# 0xda7aba5e 0xde11ca7e 0xde7a11ed 0xd1abe7e5 0xd15ab1ed 0xd15c105e 0xf007ba11 0x1501a7ed 0x0ff1c1a1 0x7ac71ca1  

# PS G:\Projects\Perl\Challenges> perl .\Hex2Words.pl 7
# 0x0be51de5 0x0c1a55e5 0x0c1a551c 0x0c011ec7 0x0dec1ded 0x0def1c17 0x0d15ea5e  

  




